### PR TITLE
use 54% dcnn, fix #32

### LINKED ISSUE #32: segment fault for dcnn
in dcnn.cpp, function dcnn_get_moves, I find board->b is NULL.

But in uct/prior.c, function uct_prior_dcnn, I find map->b->b is NOT NULL.

This is strange, is it because of thread-safety?

### PR DESCRIPTION
**Pull request by xlvector**
Fix #32 , and use 54% dcnn from http://computer-go.org/pipermail/computer-go/2015-December/008324.html.

I have test with human 2d, and can win him. And this network can also win gnugo level 10 with 2 handicap. (I only test 2 handicap, may 3 is still ok)

### CODE REVIEW

File: board.c [hunk 1]
```diff
@@ -1317,6 +1319,8 @@ board_play_in_eye(struct board *board, struct move *m, int f)
 	board_at(board, coord) = color;
 	group_t group = new_group(board, coord);
 
+	board->last_move4 = board->last_move3;
```
Comment (xlvector) [inline board.c]: dcnn need to use last 4 moves, but I find previous last_move3 and last_move4 is not set. I do not know if this modify is ok?

File: uct/prior.c [hunk 2]
```diff
@@ -113,13 +113,87 @@ print_dcnn_best_moves(struct tree_node *node, struct prior_map *map,
...
+				int offsetw = size * size + size * j + k;
+				fprintf(stderr, "board_at %d : %d\n", c, board_at(b, c));
+				data[offsetb] = (board_at(b, c) == S_WHITE ? 1.0 : 0.0);
+				data[offsetw] = (board_at(b, c) == S_BLACK ? 1.0 : 0.0);
+			}
+	}
+
+	return data;
+}
+
```
Comment (pasky) [inline uct/prior.c]: If we aren't using dcnn_get_moves_dta() anywhere, let's remove it completely. Do you agree?

File: uct/prior.c [hunk 3]
```diff
@@ -113,13 +113,87 @@ print_dcnn_best_moves(struct tree_node *node, struct prior_map *map,
         fprintf(stderr, "]\n");	
 }
 
+static float*
+dcnn2_get_moves_data(struct board *b, enum stone color)
```
Comment (pasky) [inline uct/prior.c]: I'm not that fond of moving this to uct/prior.c, but I assume you moved it out of dcnn.cpp due to the C++ issues that caused the segfaults?

Maybe we can later change it to dcnn.c + dcnn-caffe.cpp, but there's no rush for that. It's fine here imho for the time being if the above is the reason.

File: uct/prior.c (hunk 2 above)
Comment (xlvector) [inline uct/prior.c]: I agree.

**Comment by pasky** [comment]
Thanks! I'll be happy to merge it, but I'd just like @lemonsqueeze to ack it first too.

**Comment by xlvector** [comment]
Sure! @lemonsqueeze

**Comment by lemonsqueeze** [comment]
Hi,
I'm all for getting 54% dcnn support in but this seems rushed to me.
I think moving dcnn_get_moves_data() in prior.c where it doesn't belong just to work around a crash without getting to the bottom of it is just wrong. How about resolving issue #32 first ?
(btw, could anybody reproduce this crash ?)
It'd be good to play some games too just to make sure it's better than 44% one ...

**Comment by pasky** [comment]
It seems plausible to me that C++ could interpret all the intricies of
struct board differently than C for some weird reason, so I think in
principle keeping the C++ surface at the minimum makes sense.

Of course, playtesting this would be nice! :)

**Comment by xlvector** [comment]
Thanks @lemonsqueeze . I do not test them directly. But I test them against human and find human 2d can always beat 44% and can not beat 54%network. I will do more further works on this.

For crash problem, I can share the username/password of machine which can reproduce the crash. You can login to it through ssh if you have time.

**Comment by lemonsqueeze** [comment]
@xlvector ok that'd be great, or you might be able to do it yourself. If there's memory corruption we can probably find out with valgrind. You can use lemonsqueeze@gmx.com to contact me directly.

**Comment by xlvector** [comment]
Sure @lemonsqueeze . I will do some experiments and contact you directly.

**Comment by xlvector** [comment]
thanks. For performance, 

if use 44% network as black, 54% network as white, (komi = 6.5), winrate of 54% network is 58 / (58 + 52) = 53%

if use 44% network as white, 54% network as black, komi = 6.5, winrate of 54% network is 76 / (76 + 32) = 70%

I test them by:

../pachi/tools/twogtp.py --black "./pachi44 -t _120" --white "./pachi54 -t _120"

**Comment by xlvector** [comment]
@lemonsqueeze  find that, do not set MAC=1 in ubuntu can resolve segment fault problem.

But Mac users will continue get segment fault. So, its up to you close this PR or merge it.

**Comment by lemonsqueeze** [comment]
@xlvector, Now that issue #32 is cleared out we can put things in the right place.
Based on your code, proper way to do it would be something like [pachi/dcnn54][1] (this is what i use in my tree now). Are you going to clean up your branch or should i submit mine instead ?

For the test games, you are aware that `-t _120` is 2 mins per game, so less than half a second per move ? I guess multi-threaded ?

[1]: https://github.com/lemonsqueeze/pachi/tree/dcnn54

**Comment by xlvector** [comment]
Sure, you can submit your PR instead of me. So, I can close this PR?

**Comment by lemonsqueeze** [comment]
Yes, thanks for getting 54% dcnn working !

**Comment by pasky** [comment]
Superseded by PR #36.

**Pull request closed without merging**